File: skills/agents-orchestrator/scripts/backends/acp/processes.py
"""Bounded, identity-aware process-group observation and cleanup."""

import os
import signal
import subprocess
import time


def pid_alive(pid):
    if not pid:
        return False
    try:
        waited, _ = os.waitpid(int(pid), os.WNOHANG)
        if waited == int(pid):
            return False
    except ChildProcessError:
        pass
    try:
        os.kill(int(pid), 0)
        return True
    except ProcessLookupError:
        return False
    except PermissionError:
        return True


def process_group_leader_alive(pid):
    if not pid_alive(pid):
        return False
    try:
        return os.getpgid(int(pid)) == int(pid)
    except (ProcessLookupError, PermissionError):
        return False


def process_group_members(pgid):
    """Return non-zombie PIDs in a process group, or None if ps is unavailable."""
    if not pgid:
        return []
    try:
        completed = subprocess.run(
            ["ps", "-axo", "pid=,pgid=,stat="],
            stdout=subprocess.PIPE,
            stderr=subprocess.DEVNULL,
            text=True,
            check=False,
            timeout=1,
        )
    except (OSError, subprocess.TimeoutExpired):
        return None
    if completed.returncode != 0:
        return None
    members = []
    for line in completed.stdout.splitlines():
        fields = line.split(None, 2)
        if len(fields) != 3:
            continue
        try:
            pid_value, group_value = int(fields[0]), int(fields[1])
        except ValueError:
            continue
        if group_value == int(pgid) and not fields[2].upper().startswith("Z"):
            members.append(pid_value)
    return members


def process_group_alive(pgid):
    if not pgid:
        return False
    # Reap a direct child leader before checking the group. Otherwise a lone
    # zombie leader makes killpg(..., 0) look like a live execution forever.
    pid_alive(pgid)
    members = process_group_members(pgid)
    if members is not None:
        return bool(members)
    try:
        os.killpg(int(pgid), 0)
        return True
    except ProcessLookupError:
        return False
    except PermissionError:
        return True


def process_has_nonce(pid, expected_nonce):
    """Verify a persisted PID still belongs to this execution without logging env."""
    if not pid_alive(pid) or not expected_nonce:
        return False
    needle = ("AGENT_SWARM_EXECUTION_NONCE=%s" % expected_nonce).encode("utf-8")
    environ_path = "/proc/%d/environ" % int(pid)
    try:
        with open(environ_path, "rb") as stream:
            return needle + b"\0" in stream.read()
    except OSError:
        pass
    try:
        completed = subprocess.run(
            ["ps", "eww", "-p", str(int(pid)), "-o", "command="],
            stdout=subprocess.PIPE,
            stderr=subprocess.DEVNULL,
            check=False,
            timeout=1,
        )
    except (OSError, subprocess.TimeoutExpired):
        return False
    return needle in completed.stdout.split()


def wait_absent(pid, timeout=3):
    deadline = time.monotonic() + max(0, timeout)
    while time.monotonic() < deadline:
        if not pid_alive(pid):
            return True
        time.sleep(0.05)
    return not pid_alive(pid)


def wait_process_group_absent(pgid, timeout=3):
    deadline = time.monotonic() + max(0, timeout)
    while time.monotonic() < deadline:
        if not process_group_alive(pgid):
            return True
        time.sleep(0.05)
    return not process_group_alive(pgid)


def _signal_process_group(pgid, sig):
    members = process_group_members(pgid)
    try:
        os.killpg(int(pgid), sig)
        return True
    except ProcessLookupError:
        return True
    except PermissionError:
        # macOS may reject killpg after the leader exits. Fall back to the
        # already-resolved members of this verified/trusted group.
        signalled = False
        for member in members or []:
            try:
                os.kill(member, sig)
                signalled = True
            except ProcessLookupError:
                continue
            except PermissionError:
                return False
        return signalled


def terminate_process_group(pid, grace=1, *, expected_nonce=None, trusted=False):
    if not process_group_alive(pid):
        return True
    if not trusted and (
        not process_group_leader_alive(pid)
        or not process_has_nonce(pid, expected_nonce)
    ):
        return False
    if not _signal_process_group(pid, signal.SIGTERM):
        return not process_group_alive(pid)
    if wait_process_group_absent(pid, grace):
        return True
    if not _signal_process_group(pid, signal.SIGKILL):
        return not process_group_alive(pid)
    return wait_process_group_absent(pid, grace)

File: skills/agents-orchestrator/scripts/backends/acp/processes.ts
import { readFileSync } from "node:fs";

function integer(value: unknown): number | null {
  const parsed = Number(value);
  return Number.isInteger(parsed) && parsed > 0 ? parsed : null;
}

export function pidAlive(pid: unknown): boolean {
  const value = integer(pid);
  if (!value) return false;
  try {
    process.kill(value, 0);
    const status = Bun.spawnSync({ cmd: ["ps", "-o", "stat=", "-p", String(value)], stdout: "pipe", stderr: "ignore", timeout: 1000 });
    if (status.exitCode === 0 && status.stdout.toString().trim().toUpperCase().startsWith("Z")) return false;
    return true;
  }
  catch (error) { return (error as NodeJS.ErrnoException).code === "EPERM"; }
}

export function processGroupLeaderAlive(pid: unknown): boolean {
  const value = integer(pid);
  if (!value || !pidAlive(value)) return false;
  const result = Bun.spawnSync({ cmd: ["ps", "-o", "pgid=", "-p", String(value)], stdout: "pipe", stderr: "ignore", timeout: 1000 });
  return result.exitCode === 0 && Number(result.stdout.toString().trim()) === value;
}

export function processGroupMembers(pgid: unknown): number[] | null {
  const value = integer(pgid);
  if (!value) return [];
  const result = Bun.spawnSync({ cmd: ["ps", "-axo", "pid=,pgid=,stat="], stdout: "pipe", stderr: "ignore", timeout: 1000 });
  if (result.exitCode !== 0) return null;
  const members: number[] = [];
  for (const line of result.stdout.toString().split("\n")) {
    const match = /^\s*(\d+)\s+(\d+)\s+(\S+)/u.exec(line);
    if (match && Number(match[2]) === value && !match[3]!.toUpperCase().startsWith("Z")) members.push(Number(match[1]));
  }
  return members;
}

export function processGroupAlive(pgid: unknown): boolean {
  const value = integer(pgid);
  if (!value) return false;
  const members = processGroupMembers(value);
  if (members !== null) return members.length > 0;
  try { process.kill(-value, 0); return true; }
  catch (error) { return (error as NodeJS.ErrnoException).code === "EPERM"; }
}

export function processHasNonce(pid: unknown, expectedNonce: unknown): boolean {
  const value = integer(pid);
  if (!value || !pidAlive(value) || typeof expectedNonce !== "string" || !expectedNonce) return false;
  const needle = `AGENTS_ORCHESTRATOR_EXECUTION_NONCE=${expectedNonce}`;
  try { return readFileSync(`/proc/${value}/environ`).toString().split("\0").includes(needle); } catch { /* macOS */ }
  const result = Bun.spawnSync({ cmd: ["ps", "eww", "-p", String(value), "-o", "command="], stdout: "pipe", stderr: "ignore", timeout: 1000 });
  return result.exitCode === 0 && result.stdout.toString().split(/\s/u).includes(needle);
}

function waitUntil(check: () => boolean, timeoutSeconds: number): boolean {
  const deadline = performance.now() + Math.max(0, timeoutSeconds) * 1000;
  while (performance.now() < deadline) {
    if (check()) return true;
    Bun.sleepSync(50);
  }
  return check();
}

export function waitAbsent(pid: unknown, timeoutSeconds = 3): boolean {
  return waitUntil(() => !pidAlive(pid), timeoutSeconds);
}

export function waitProcessGroupAbsent(pgid: unknown, timeoutSeconds = 3): boolean {
  return waitUntil(() => !processGroupAlive(pgid), timeoutSeconds);
}

function signalGroup(pgid: number, signal: NodeJS.Signals): boolean {
  const members = processGroupMembers(pgid);
  try { process.kill(-pgid, signal); return true; }
  catch (error) {
    if ((error as NodeJS.ErrnoException).code === "ESRCH") return true;
    if ((error as NodeJS.ErrnoException).code !== "EPERM") return false;
    let signalled = false;
    for (const member of members ?? []) {
      try { process.kill(member, signal); signalled = true; }
      catch (memberError) { if ((memberError as NodeJS.ErrnoException).code !== "ESRCH") return false; }
    }
    return signalled;
  }
}

export function terminateProcessGroup(
  pid: unknown,
  options: { graceSeconds?: number; expectedNonce?: string | null; trusted?: boolean } = {},
): boolean {
  const value = integer(pid);
  if (!value || !processGroupAlive(value)) return true;
  if (!options.trusted && (!processGroupLeaderAlive(value) || !processHasNonce(value, options.expectedNonce))) return false;
  if (!signalGroup(value, "SIGTERM")) return !processGroupAlive(value);
  if (waitProcessGroupAbsent(value, options.graceSeconds ?? 1)) return true;
  if (!signalGroup(value, "SIGKILL")) return !processGroupAlive(value);
  return waitProcessGroupAbsent(value, options.graceSeconds ?? 1);
}
